internal/guest/storage/scsi/detect.go

```diff
@@ -0,0 +1,168 @@
+		if err != nil {
+			return 0, 0, fmt.Errorf("invalid scsi device name: %s", name)
+		}
+		switch i {
+		case 0:
+			controller = uint8(n)
+		case 3:
+			lun = uint8(n)
+		}
+	}
+	return
+}
+
+// detectNewScsiDevice looks for all the SCSI devices attached on the host compares them with the
+// previously known devices and returns the path of that device which can be used for mounting it.
+func detectNewScsiDevice(ctx context.Context) (uint8, uint8, error) {
+	// all scsi devices show up under /sys/bus/scsi/devices
+	// The devices are named as 0:0:0:0, 0:0:0:1, 1:0:0:0 etc.
+	// The naming format is as follows:
+	// <host controller no.>:<bus>:<target>:<LUN>
+	// (Section 3.1 from https://tldp.org/HOWTO/html_single/SCSI-2.4-HOWTO/)
+
+	for {
+		devices, err := os.ReadDir(scsiDevicesPath)
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to read entries from %s: %w", scsiDevicesPath, err)
+		}
+
+		for _, dev := range devices {
+			name := path.Base(dev.Name())
+			newController, newLUN, err := parseScsiDeviceName(name)
+			if err != nil {
+				continue
+			}
+			if _, ok := knownScsiDevices[name]; !ok {
+				knownScsiDevices[name] = true
+				return newController, newLUN, nil
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return 0, 0, fmt.Errorf("context timed out waiting for SCSI device")
+		default:
+			time.Sleep(time.Millisecond * 10)
+			continue
+		}
+
+	}
+}
+
+func removeKnownDevice(controller, LUN uint8) {
+	name := fmt.Sprintf("%d:0:0:%d", controller, LUN)
+	delete(knownScsiDevices, name)
+}
+
+// recordActualSCSIMapping keeps a mapping of SCSI devices that are mapped at different
+// controller/LUN number than expected.
+func recordActualSCSIMapping(origController, origLUN, actualController, actualLUN uint8) error {
+	origPath := fmt.Sprintf("%d:0:0:%d", origController, origLUN)
+	actualPath := fmt.Sprintf("%d:0:0:%d", actualController, actualLUN)
+	if origPath != actualPath {
+		logrus.Debugf("SCSI disk expected at %s, actually attached to: %s", origPath, actualPath)
+	}
+	if _, ok := actualMappings[origPath]; !ok {
+		actualMappings[origPath] = actualPath
+	} else {
+		return fmt.Errorf("double mapping of scsi device %s", origPath)
+	}
+	return nil
+}
+
+func getActualMapping(origController, origLUN uint8) (actualController, actualLUN uint8) {
+	origPath := fmt.Sprintf("%d:0:0:%d", origController, origLUN)
+	if actualMapping, ok := actualMappings[origPath]; ok {
+		// should never return error
+		actualController, actualLUN, _ = parseScsiDeviceName(actualMapping)
+	} else {
+		return origController, origLUN
+	}
+	return
+}
+
+func removeActualMapping(origController, origLUN uint8) (actualController, actualLUN uint8) {
+	origPath := fmt.Sprintf("%d:0:0:%d", origController, origLUN)
+	actualController, actualLUN = getActualMapping(origController, origLUN)
+	delete(actualMappings, origPath)
+	return
+}
+
+// Mount is the wrapper on top of the actual Mount call to detect the correct controller and LUN numbers.
+func Mount(
```

Comment:
The rest of the file makes sense to be in detect.go, but maybe the wrappers should live in scsi.go with what they're wrapping (`mount` and friends)? For me at least, I'd expect to find `Mount` in scsi.go (or if we had a mount.go file in the pkg)